### PR TITLE
fix stubGenerator

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #23725 .

### Summary of Changes

renamed file `libraries/src/Filesystem/Support/Stringcontroller.php`  to `libraries/src/Filesystem/Support/StringController.php`

### Testing Instructions
run: php build/stubGenerator.php

### Expected result
file written

### Actual result
Error: Failed to start application: Class Joomla\CMS\Filesystem\Support\StringController does not exist

### CODE REVIEW

**Comment by Bakual** [comment]
You renamed it to `StringcController.php`. There is a `c` to much in there 😄

**Comment by alikon** [comment]
yes ,fixed thanks

**Comment by brianteeman** [comment]
Its not that easy :(

Changing filename case on a windows system on upgrade is a PIA

We've had this before although right now I forget exactly what the file was - perhaps @mbabker remembers

**Comment by Bakual** [comment]
Brian is right.
Windows is actually fine. It's Linux where you get an issue because they count as different files and you end up with two files.
Now we can add the old file to be deleted into the admin script, which would work on Linux. But it would fail on Windows since it would delete the new file (it ignores case).
So there is no simple solution. You would need a script which checks OS or if the file is there twice and only then delete it.
I think last time we had that issue we just moved the file to another location, then it's simple.

**Comment by mbabker** [comment]
> I think last time we had that issue we just moved the file to another location, then it's simple.

Which means deprecating yet another class and renaming it something else when talking about any file that contains a PHP class following an autoloading convention.

**Comment by mbabker** [comment]
> We've had this before although right now I forget exactly what the file was - perhaps @mbabker remembers

IIRC that was when the JClassLoader file was at the root of the `libraries` directory with a camel cased filename or something like that and we just shoved it into the `libraries/cms` directory following the normal file convention.

**Comment by Bakual** [comment]
> Which means deprecating yet another class and renaming it something else when talking about any file that contains a PHP class following an autoloading convention.

Yep exactly. Not a nice way, but the easiest one.

**Comment by HLeithner** [comment]
We could fix the deletion process at https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_admin/script.php#L2257-L2263

if we use realpath to get the real filename we also get the correct case

something like this (just a idea)
```php
if (JFile::exists(JPATH_ROOT . $file) && realpath(JPATH_ROOT . $file) === JPATH_ROOT . $file && !JFile::delete(JPATH_ROOT . $file))
```

**Comment by alikon** [comment]
...could be

**Comment by HLeithner** [comment]
Maybe I didn't explained it correct but the reason to use realpath is to match the filename with case on disk with the filename we want to delete.

if you use this code you gain nothing, you have to compare the result auf realpath with the filestring we want to delete.
JFile::exists uses is_file and this is case insensitive, the same is true for unlink used by JFile::delete

btw I think JFile::exists can be omitted because JFile::delete does this anyway

File: administrator/components/com_admin/script.php [hunk 1]
```diff
@@ -2256,7 +2258,7 @@ public function deleteUnexistingFiles()
 
 		foreach ($files as $file)
 		{
-			if (JFile::exists(JPATH_ROOT . $file) && !JFile::delete(JPATH_ROOT . $file))
+			if (JFile::exists(realpath(JPATH_ROOT . $file)) && !JFile::delete(realpath(JPATH_ROOT . $file)))
```
Comment (HLeithner) [inline administrator/components/com_admin/script.php]: Change this to
```suggestion
			if (realpath(JPATH_ROOT . $file) === JPATH_ROOT . $file && !JFile::delete(JPATH_ROOT . $file))
```

**Comment by alikon** [comment]
no it's me on the train  :smile:

**Comment by HLeithner** [comment]
hmm this will not work if your webserver access the site with a symlink... So we need more magic here

**Comment by HLeithner** [comment]
@mbabker do you have a idea? Maybe testing realpath(JPATH) and realpath($PATH) seperated?

**Comment by ghost** [comment]
@Orgoth please retest (test comment in #23725 was in february) and mark your test as successfully > https://docs.joomla.org/Testing_Joomla!_patches#Recording_test_results

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on d6f533e5c6b4d37dd9dd20180640ec6d2611719c<br /><br />

```
isley@isley:~/repos/test/joomla$ php build/stubGenerator.php
Stubs file written
```
File was created and looks good so far.
```
-rw-rw-r--  1 isley isley  54439 Apr 24 14:15 stubs.php
```

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23762">issues.joomla.org/tracker/joomla-cms/23762</a>.</sub>

File: administrator/components/com_admin/script.php [hunk 2]
```diff
@@ -1983,6 +1983,8 @@ public function deleteUnexistingFiles()
 			'/libraries/joomla/filesystem/wrapper/path.php',
 			'/libraries/src/Mail/language/phpmailer.lang-joomla.php',
 			'/plugins/captcha/recaptcha/recaptchalib.php',
+			// Joomla! 3.9.4
+			'/libraries/src/Filesystem/Support/Stringcontroller.php',
```
Comment (HLeithner) [inline administrator/components/com_admin/script.php]: please revert this change and the StringController.php renaming

**Comment by gwsdesk** [comment]
I have tested this item :white_check_mark: successfully on 8ff8935394f353bdcb93ad97646e3fbf3a04105b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23762">issues.joomla.org/tracker/joomla-cms/23762</a>.</sub>

**Comment by gwsdesk** [comment]
@Orgoth can you please post your successful test at the Tracker? We can merge it than since we have 2 successful tests

**Comment by HLeithner** [comment]
It would be great if testers add the operating system to the test node, because this could lead to problems on different Systems.

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 8ff8935394f353bdcb93ad97646e3fbf3a04105b<br /><br />### Details:
System	Kubuntu 18.04 LTS
PHP Built On	Linux isley.lc 4.15.0-47-generic #50-Ubuntu SMP Wed Mar 13 10:44:52 UTC 2019 x86_64
Database Type	mysql
Database Version	8.0.15-5
Database Collation	utf8_general_ci
Database Connection Collation	utf8mb4_0900_ai_ci
PHP Version	7.1.28-1+ubuntu18.04.1+deb.sury.org+3
Web Server	Apache/2.4.39 (Ubuntu)
WebServer to PHP Interface	fpm-fcgi<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23762">issues.joomla.org/tracker/joomla-cms/23762</a>.</sub>

**Comment by HLeithner** [comment]
A windows test would be awesome.

**Comment by ghost** [comment]
I am very sorry, but I only develop under Linux.
Maybe @gwsdesk has the possibility to test under Windows.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23762">issues.joomla.org/tracker/joomla-cms/23762</a>.</sub>

**Comment by brianteeman** [comment]
stubgenerator runs fine no matter the case of that file on windows

**Comment by HLeithner** [comment]
thx @brianteeman did you tested it with already existing files, lower and uppercase before and after upgrade?

**Comment by brianteeman** [comment]
I ran the script before renaming the file and after renaming the file - it ran correctly in both cases

**Comment by bembelimen** [comment]
I have tested this item :red_circle: unsuccessfully on a0538dc6550be26675056e029b523c0c1ee0c2f7<br /><br />In windows this PR breaks the whole deletion process, because ```realpath``` converts all slashes to ```\```, so it's always not true (and the JFile::delete will never be called).

Additional there will not be 2 files in windows, so the deletion process (if it works) would delete the whole file without replacement.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23762">issues.joomla.org/tracker/joomla-cms/23762</a>.</sub>

**Comment by HLeithner** [comment]
Ok so we have to find another solution, I would suggest an own function for renaming files, before calling the delete function.
